lib/runtime/index.js

```diff
@@ -160,3 +237,96 @@ function runSetters(valueToPassThrough, names) {
 function moduleMakeNsSetter(includeDefault) {
   return this.exportAs(includeDefault ? "*+" : "*");
 }
+
+function wrapAsync(body, options) {
+  var module = this;
+  var entry = Entry.getOrCreate(module.id, module);
+  entry.hasTLA = options.async;
+
+  // Caches the result for waitForDeps. This can be:
+  // - undefined before waitForDeps was called
+  // - null if there are no deps to wait for
+  // - a promise if there are deps to wait for
+  var waitForDepsResult = undefined;
+
+  body.call(options.self, module, function waitForDeps() {
+    // waitForDeps is called twice, once to check if the module should wait
+    // and again to await the promise
+    if (waitForDepsResult === undefined) {
+
+      // This code should run immediately after all deps are synchronously required
+      // By setting entry.asyncEvaluation here, we avoid issues of dead locks from
+      // circular deps - this module's deps won't be able to add
+      // an _onEvaluating hook to this module
+      entry.setAsyncEvaluation();
```

Comment:
This comment might be misleading. The primary reason we do this here is because this is when the spec requires it to be done. The comment might be from when we were implementing an older version of the spec.